### PR TITLE
Feature/before uncompiled hook

### PR DESCRIPTION
**Pull request by t-b**
Created from http://stash.corp.alleninstitute.org/projects/ENG/repos/mies-igor/pull-requests/1119/overview.

### CODE REVIEW

**Comment by t-b** [comment]
The errors on the hardware tests here are real, running `runtest("UTF_.*", testcase="PS_RB10", enableregexp=1)` fails with

```
  Start of test "Unnamed"
  Entering test suite "UTF_PatchSeqRheobase.ipf"
  Entering test case "PatchSeqTestRheobase#PS_RB10:ITC"
  Available number of specified ITC devices = 1
  ITC Device opened, returned deviceID is 0.
  (ITC18USB_dev_0): Could not find final DAScale value from one of the previous analysis functions.
  TEST OVERRIDE ACTIVE
  (ITC18USB_dev_0): Could not find final DAScale value from one of the previous analysis functions.
  !!! Assertion FAILED !!!
  Message: "Async frame work already running"
  !!! Assertion FAILED !!!
  Message: "Async frame work is not running"
```

The same test passes with IP8.
@timjarsky Should we invest some more time to fix it for IP7 or should we go IP8-only now?

**Comment by timjarsky** [comment]
> @timjarsky Should we invest some more time to fix it for IP7 or should we go IP8-only now?

For now, why don't we just skip this test for IP7? I'd like to make sure end users have IP8 in place before transitioning completely away from IP7.

**Comment by t-b** [comment]
@timjarsky It is not about only one test. The general approach with starting/stopping the async framework seems to be broken. This test just exposes it.

**Comment by timjarsky** [comment]
> @timjarsky It is not about only one test. The general approach with starting/stopping the async framework seems to be broken. This test just exposes it.

When was this test implemented?

**Comment by t-b** [comment]
@MichaelHuth Conclusion was that this needs to be fixed for IP7.

**Comment by t-b** [comment]
`Async_Start`/`Async_Stop` should be made idempotent using `ThreadGroupWait` tgID check.

**Comment by MichaelHuth** [comment]
I have tested `UTF_PatchSeqRheobase` in IP7 and it doesn't show errors any more.

File: Packages/MIES/MIES_Async.ipf [hunk 1]
```diff
@@ -37,20 +37,23 @@ static StrConstant ASYNC_ERRORMSG_STR = "errmsg"
...
-		// framework already running
+	DFREF dfr = GetAsyncHomeDF()
+	if(IsASYNCRunning())
 		return 0
 	endif
 
 	ASSERT(numThreads >= 1 && numThreads <= ASYNC_MAX_THREADS, "numThread must be > 0 and <= " + num2str(ASYNC_MAX_THREADS))
 	disableTask = ParamIsDefault(disableTask) ? 0 : !!disableTask
 
+
```
Comment (t-b) [inline Packages/MIES/MIES_Async.ipf]: unexpected newline

File: Packages/MIES/MIES_Async.ipf [hunk 2]
```diff
@@ -1031,3 +1036,24 @@ static Function/WAVE GetWorkloadOrder(dfr)
 	SetDimLabel 1, 1, $"orderGlobal", wv
 	return wv
 End
+
+/// @brief returns 1 if ASYNC framework is running, 0 otherwise
+static Function IsASYNCRunning()
```
Comment (t-b) [inline Packages/MIES/MIES_Async.ipf]: Should use `ASYNC_` name prefix.

File: Packages/MIES/MIES_Async.ipf [hunk 3]
```diff
@@ -1031,3 +1036,24 @@ static Function/WAVE GetWorkloadOrder(dfr)
...
+
+	variable waitResult, err, doe
+
+	NVAR tgID = $GetThreadGroupID()
+	try
+		DebuggerOptions
```
Comment (t-b) [inline Packages/MIES/MIES_Async.ipf]: Please use the dedicated helpers: `ResetDebugOnError` and `DisableDebugOnError`.

File: Packages/MIES/MIES_Async.ipf (hunk 1 above)
Comment (MichaelHuth) [reply Packages/MIES/MIES_Async.ipf]: I was still in the progress of resorting/cleaning up the commits.

File: Packages/MIES/MIES_Async.ipf (hunk 2 above)
Comment (MichaelHuth) [reply Packages/MIES/MIES_Async.ipf]: done

File: Packages/MIES/MIES_Async.ipf (hunk 3 above)
Comment (MichaelHuth) [reply Packages/MIES/MIES_Async.ipf]: done

**Comment by t-b** [comment]
@timjarsky Tests, incuding the hardware tests, pass.

**Pull request merged**